src/NBitcoin/ChainedHeader.cs

```diff
@@ -286,19 +286,24 @@ public ChainedHeader FindAncestorOrSelf(ChainedHeader chainedHeader)
         }
 
         /// <summary>
-        /// Finds the ancestor of this entry in the chain that matches the block hash given.
+        /// Finds the ancestor of this entry in the chain that matches the block hash.
+        /// It will not search lower than the optional height parameter.
         /// </summary>
         /// <param name="blockHash">The block hash to search for.</param>
-        /// <returns>The ancestor of this chain that matches the block hash.</returns>
-        public ChainedHeader FindAncestorOrSelf(uint256 blockHash)
+        /// <param name="height">Optional height to restrict the search to.</param>
+        /// <returns>The ancestor of this chain that matches the block hash, or null if it was not found.</returns>
+        public ChainedHeader FindAncestorOrSelf(uint256 blockHash, int height = 0)
         {
             ChainedHeader currentBlock = this;
-            while ((currentBlock != null) && (currentBlock.HashBlock != blockHash))
+            while ((currentBlock != null) && (currentBlock.Height > height))
             {
+                if (currentBlock.HashBlock == blockHash)
```

Comment:
Supposedly you can save an additional `if` statement  if you replace the condition `currentBlock.HashBlock == blockHash` with `currentBlock.Height > height`

And in the if do this
 ```
if(currentBlock.Height > height)
{
    currentBlock = null;
    break;
}

then the return is just `currentBlock`

```

But I am happy with this as is as well.